### ISSUE TITLE
[Feature Request] Any plan to support fp8 type at latest torch version?

### ISSUE THREAD

**Issue by MoFHeka** [graphcore-research/unit-scaling#45]
The latest torch has been supported the type e5m2 and e4m3.

**Comment by thecharlieblake** [comment]
Thanks for the request - doesn't seem too difficult to add and definitely worthwhile, I'll try and get around to it this week (if you want it sooner, feel free to open a PR)

**Comment by thecharlieblake** [comment]
@MoFHeka @lin72h I've had a look into this and it seems like the PyTorch 2.1.0 support for FP8 is still too limited to make this work for unit scaling. There are two key issues here:

1. It doesn't support matmuls where one operand is FP8 E5 and the other is FP8 E4. This is a problem as to compute gradients with respect to activations we usually multiply E4 weights with E5 gradients.
2. Matmuls where both input operands are in FP8, currently output the result in FP8. This isn't really our desired behaviour. The literature tends to assume the output is in higher-precision (you have to accumulate in higher-precision, so why not output in it?), e.g. FP16, which is essential if your next operation doesn't work in FP8 (e.g. within self-attention). If we adopted the current PyTorch approach, we'd have to do _everything_ in FP8, not just the matmuls. It's not clear this would work...

Implementations for specific devices (GPU, IPU, etc) may have their own approach which fixes these things. If anyone wants to add hardware-specific FP8 support, feel free to send a PR!

**Comment by MoFHeka** [comment]
@thecharlieblake 
In my understanding, if too much operation is spent on the conversion between FP8 and FP16, then the computing speed and bandwidth advantages of FP8 will be eroded.
In addition, now OpenAI triton also supports pytorch FP8 format for custom operators, will it be more conducive to FP8 development?

**Comment by thecharlieblake** [comment]
In theory yes, if that cast is too expensive it could erode the benefit - but in practice I'd expect any hardware provider that supports FP8 arithmetic to also have a fast FP8 cast supported by their hardware, so in practice I doubt that will be too big an issue.

I'd imagine Triton support would help FP8 development, yes - although only on hardware that's Triton-friendly (currently Nvidia and increasingly AMD GPUs).

**Comment by MoFHeka** [comment]
What does hardware support mean? Do you mean that during training, the hardware computing unit can directly load FP16 in HBM as FP8 into the cache for calculation, instead of applying for more video memory to store the corresponding FP8 parameters of FP16 in HBM?
